sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/SymbolicExecution/Constraints/SymbolicValueConstraints.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using SonarAnalyzer.Common;
+
+namespace SonarAnalyzer.Helpers.FlowAnalysis.Common
+{
+    public class SymbolicValueConstraints
+    {
+        private readonly Dictionary<Type, SymbolicValueConstraint> constraints
+            = new Dictionary<Type, SymbolicValueConstraint>();
+
+        private readonly int hashCode;
+
+        public SymbolicValueConstraints(SymbolicValueConstraint constraint)
+        {
+            SetConstraint(constraint, this.constraints);
+            hashCode = ComputeHashcode();
+        }
+
+        private SymbolicValueConstraints(Dictionary<Type, SymbolicValueConstraint> constraints)
+        {
+            this.constraints = constraints;
+            hashCode = ComputeHashcode();
+        }
+
+        internal IEnumerable<SymbolicValueConstraint> GetConstraints() => constraints.Values;
+
+        internal virtual SymbolicValueConstraints SetConstraint(SymbolicValueConstraint constraint)
+        {
+            var constraintsCopy = new Dictionary<Type, SymbolicValueConstraint>(constraints);
+            SetConstraint(constraint, constraintsCopy);
+
+            return new SymbolicValueConstraints(constraintsCopy);
+        }
+
+        private static void SetConstraint(SymbolicValueConstraint constraint,
+            Dictionary<Type, SymbolicValueConstraint> constraints)
+        {
+            constraints[constraint.GetType()] = constraint;
+
+            if (constraint is BoolConstraint ||
+                constraint is DisposableConstraint)
```

Comment:
+1 We need to somehow avoid making changes here when adding new constraints. Perhaps we should reintroduce the `Implies()` method as a property that will return the constraints that are implied when adding this one. For example, Bool and Disposable constraints should imply NotNull and basically you will have to just add the constraint and all implied constraints.